### ISSUE TITLE
Async methods are misleading

### ISSUE THREAD

**Issue by mj1856** [zzzprojects/EntityFramework-Plus#131]
Hi,

Came across your library through [this StackOverflow question](http://stackoverflow.com/questions/42422656/entity-framework-doesnt-provide-deleteasync-or-addasync-why).

Please watch [this short MSDN video](https://channel9.msdn.com/Series/Three-Essential-Tips-for-Async/Async-Library-Methods-Shouldn-t-Lie) which explains why things like your implementation of `DeleteAsync` is misleading.

```
public static Task<int> DeleteAsync<T>(this IQueryable<T> query, CancellationToken cancellationToken = default(CancellationToken)) where T : class
{
    return Task.Run(() => query.Delete(null), cancellationToken);
}
```

In short, simply wrapping a synchronous operation with `Task.Run` does not actually make the method asynchronous.  it just adds overhead for dealing with the task, thread switching, etc.

**Comment by JonathanMagnan** [comment]
Hello @mj1856 ,

Thank you for sharing the video. I will look at it this weekend.

Yes, I'm aware of the issue about `Task.Run` & Async. We already fixed Query Cache feature to make it truly Async (Issue #98)

Unfortunately, time to support this library is currently missing. I will try to look on it this weekend, but I cannot make any promise.

Best Regards,

Jonathan

**Comment by neurospeech** [comment]
@mj1856 I think the statement 

> In short, simply wrapping a synchronous operation with Task.Run does not actually make the method asynchronous. it just adds overhead for dealing with the task, thread switching, etc.

 is misunderstood mostly.

Though it does not make it async, but a workaround is not that bad as this feature can be added in future without you having to change sync method to async. People arguing that async method should not be exposed if it isn't truly async, do not understand how libraries are built and released.

BatchDelete already offers great performance without having to load entities in memory, think Sql Azure etc, you are saving bandwidth and you are reducing deadlocks by saving time required for sql to finish operations.

If you want your app to use async, then you can continue using fake `Task.Run` async, and in future upgrade to new version with actual async will not require you to rewrite your code. And I guess author knows this very well.

**Comment by JonathanMagnan** [comment]
Hello @neurospeech ,

I have learned last year about `Task.Run` not being truly async but didn't make any modification since I still argue with myself ;)

As you said, by using some feature like Batch Delete, you already save a lot of time and bandwidth so this additional thread is for sure not a bottleneck compared to a lot of other things that can make the application slower.

From my opinion, when working with database, async method should probably be avoided. It causes more performance problem than it solve.

See: [EF Async takes ten times as long to complete](https://stackoverflow.com/a/28619983/5619143)

Having a database with some large field is very common!

Honestly, if it was only me, I would only use `Task.Run` for everything that needs to read from the database. There is more chance you get affected by the `Async` performance issue than by the additional thread taken by `Task.Run`

Best Regards,

Jonathan

**Comment by Bosch-Eli-Black** [comment]
We recently got bitten by by that EF Async bug, too (https://stackoverflow.com/questions/28543293/entity-framework-async-operation-takes-ten-times-as-long-to-complete/28619983#28619983)! 🙂
